docs/reference.md

```diff
@@ -131,3 +131,64 @@ dedicated test-suite, consisting of several tests exercising various runtime
 behaviors, for each StableHLO Op. The tests can be found
 [here](https://github.com/openxla/stablehlo/tree/main/stablehlo/tests/) (e.g.
 interpret\_\*.mlir).
+
+### Testing guidelines
+
+**(G1) Do we need to test for all the supported types for every op?**
+
+We can use a combination of following rules to decide it:
+
+  1. While implementing an op, if there exists code in the corresponding `eval`
+     function to handle a particular type, then it is imperative to have test(s)
+     to cover for that type. As an example, for `add` op, there is exclusive
+     code to handle integer, boolean, floating-point, and complex types, and
+     hence we need one test for each category of types.
+  2. If a set of types is handled uniformly in the corresponding
+     `eval` function, then a single test for all those types should be
+     sufficient. As an example, for `add` op, all the variants of integer
+     types (`si4`, `u4`, `si8`, `u8` and so on) are handled alike using
+     `llvm::APInt` APIs, and hence we can skip adding tests for each of those
+     variants, and instead, add a single representative test. To avoid ambiguity
+     in selecting the representative, we should use the following guidelines:
+
+       * If all the types, handled uniformaly, have the same primitive type
+         (i.e., if all are integer, or floating-point, or complex types), then
+         choose the one with maximum bit-width.
+       * If all the types, handled uniformaly, have a mix of primitive types,
+         then choose the one with the following primitive type, in decreasing
+         order of preference: integer, floating-point, boolean, complex.
+
+**(G2) How about adding tests dedicated for testing the interpreter infrastructure?**
+
+The interpreter infrastructure is mostly straightforward and can be added it in
+our trust base. The only non-trivial part is how various types are packed into
+and unpacked from the underlying interpreter storage. As discussed in (G1), we
+will be testing only those types of an op which are handled differently. With
+that is possible that the packing/un-packing code, corresponding to different
+variants of integer/floating-point types, might not get fully covered during
+testing.
+
+To ensure that we can chose an op, like `add`, which supports all the StableHLO
+element types and write exhaustive tests. Fortunately, we already have it.
+
+**(G3) If the implementation of an op depends other ops, should be write tests for the latter?**
+
+No. For example, the implementation of `batch_norm_grad` can be based on
+`divide`, `subtract`, `multiply` and others, we should avoid testing the latter
+ops while testing the former.
```

Comment:
This seems right, I have a question - We would still be able to interpret `batch_norm_grad` right? As in it would be lowered to other ops and those ops would be interpreted and give a result?

Reply:
You that is right! Given that the interpreter needs to be a readable implementation of the spec, it is expected that the implementation of a batch* op follow the from the evaluation of  the decomposed ops.